File: 121_best_time_to_buy_and_sell_stock/121_best_time_to_buy_and_sell_stock.cpp
/*************************************************************************
	> File Name: 121_best_time_to_buy_and_sell_stock.cpp
	> Author: Tingjian Lau
	> Mail: tjliu@mail.ustc.edu.cn
	> Created Time: 2016/05/16
 ************************************************************************/

class Solution {
public:
    int maxProfit(vector<int>& prices) {
       int maxProfit = 0;
       int minPrice = INT_MAX;
       int len = prices.size();
       
       for(int i=0; i<len; ++i){
           minPrice = min(minPrice, prices[i]);
           maxProfit = max(maxProfit, prices[i] - minPrice);
       }
       
       return maxProfit;
    }
};

File: 121_best_time_to_buy_and_sell_stock/121_best_time_to_buy_and_sell_stock.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

##########################################################################
#	> File Name: 121_best_time_to_buy_and_sell_stock.py
#	> Author: Tingjian Lau
#	> Mail: tjliu@mail.ustc.edu.cn
#	> Created Time: 2016/05/16
#########################################################################

class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        maxProfit, minPrice = 0, sys.maxint # 整数的最大值
        
        for i in range(0, len(prices)):
            minPrice = min(minPrice, prices[i])
            maxProfit = max(maxProfit, prices[i] - minPrice);
            
        return maxProfit
